JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -10,91 +10,339 @@
+		try
+		{
+			for (int result = in.read(); result != -1; result = in.read()) 
+			{
+				buf.write((byte) result);
+			}
+		}catch(Exception e)
+		{
+			e.printStackTrace();
+		}
+		return buf;
+	}
+	// get and set tokens from IDP 
+	private static void GetAndSetTokens(Map<String, String> formData) throws Exception
+	{
+		try {
+			String postOpts = getEncodedParamsString(formData);
+			byte[] postData = postOpts.getBytes("UTF-8");
+			int postDataLength = postData.length;
+			URL url = new URL(prop.getProperty("TokenUrl"));
+			HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+	        try {
+				connection.setDoOutput(true);
+				connection.setUseCaches(false);
+				connection.setRequestMethod("POST");
+				connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+				connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+				connection.setRequestProperty("Accept", "application/json");
+				connection.getOutputStream().write(postData);
+				BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
+				ByteArrayOutputStream buf = new ByteArrayOutputStream();
+				readResult(in, buf);
+				String res = buf.toString("UTF-8");
+				JsonElement jElement = JsonParser.parseString(res);
+				JsonObject jObject = jElement.getAsJsonObject();
+				// set Tokens as System Properties - new values to access_token and refresh_token
+				String accessToken = jObject.has("access_token") ? jObject.get("access_token").getAsString() : null;
+				String refreshToken = jObject.has("refresh_token") ? jObject.get("refresh_token").getAsString() : null;           
+				if (null == accessToken || null == refreshToken)
+				{
+					throw new Exception("Access/refresh token is null, Please verify the config.properties for proper inputs.");
+				}
+				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, accessToken);
+				System.setProperty(OAUTH_REFRESH_TOKEN_VAR_STRING, refreshToken);
+	        }catch(UnsupportedEncodingException uee)
+		{
+	        	uee.printStackTrace();
+	        } catch (Exception e) 
+	        {
+				String res = "";
+				try 
+				{
+					BufferedInputStream in = new BufferedInputStream(connection.getErrorStream());
+					ByteArrayOutputStream buf = new ByteArrayOutputStream();
+					readResult(in, buf);
+					res = buf.toString("UTF-8");
+				} catch (Exception ex) 
+				{
+					//Improper setup. Passes in SF, fails in Devjail. Skip when this happens, but print the error.
+					ex.printStackTrace();
+				}
+				throw e;
+	        } finally 
+	        {
+	        	connection.disconnect();
+	        }
+		}catch (IOException unreportedex) 
+		{
+			unreportedex.printStackTrace();
+		}
+	}
+	// if no access token is set, obtains and sets first access and refresh tokens
+	// uses the password grant flow
+	private static void EnsureAccessToken() throws Exception 
+	{
+		try 
```

Comment:
The added logic in this function is a little off. It should go like this:
If there is an access token, then it needs to be set, else
If there is not an access token and there is no refresh token, then get new tokens by password grant.
If there is not an access token but there is a refresh token, get new tokens by refresh grant.
If that refresh token has expired, then catch the exception and get tokens by password grant.

This logic is in ADO.NET sample application.

Reply:
-> Code changes as per suggestion are completed. 
-> EnsureAccesstoken and connectToDatabase code changed as per the ADO.Net Example.